In [2]:
import pandas as pd

# Шаг 1

In [3]:
data = pd.read_csv("proekt_1.csv")

In [4]:
data.head()

children  days_employed  dob_years education  education_id  \
0         1   -8437.673028         42    высшее             0   
1         1   -4024.803754         36   среднее             1   
2         0   -5623.422610         33   Среднее             1   
3         3   -4124.747207         32   среднее             1   
4         0  340266.072047         53   среднее             1   

      family_status  family_status_id gender income_type  debt   total_income  \
0   женат / замужем                 0      F   сотрудник     0  253875.639453   
1   женат / замужем                 0      F   сотрудник     0  112080.014102   
2   женат / замужем                 0      M   сотрудник     0  145885.952297   
3   женат / замужем                 0      M   сотрудник     0  267628.550329   
4  гражданский брак                 1      F   пенсионер     0  158616.077870   

                      purpose  
0               покупка жилья  
1     приобретение автомобиля  
2               покупка жилья  
3  дополнительное образование  
4             сыграть свадьбу

# Шаг 2

**Посмторим есть ли пропуски в таблице:**

In [5]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


**Посмотрим в каких строках присутсвуют пропуски и сделаем выводы**

In [6]:
data.loc[data["total_income"].isna()]

children  days_employed  dob_years education  education_id  \
12            0            NaN         65   среднее             1   
26            0            NaN         41   среднее             1   
29            0            NaN         63   среднее             1   
41            0            NaN         50   среднее             1   
55            0            NaN         54   среднее             1   
...         ...            ...        ...       ...           ...   
21489         2            NaN         47   Среднее             1   
21495         1            NaN         50   среднее             1   
21497         0            NaN         48    ВЫСШЕЕ             0   
21502         1            NaN         42   среднее             1   
21510         2            NaN         28   среднее             1   

               family_status  family_status_id gender  income_type  debt  \
12          гражданский брак                 1      M    пенсионер     0   
26           женат / замужем                 0      M  госслужащий     0   
29     Не женат / не замужем                 4      F    пенсионер     0   
41           женат / замужем                 0      F  госслужащий     0   
55          гражданский брак                 1      F    пенсионер     1   
...                      ...               ...    ...          ...   ...   
21489        женат / замужем                 0      M    компаньон     0   
21495       гражданский брак                 1      F    сотрудник     0   
21497        женат / замужем                 0      F    компаньон     0   
21502        женат / замужем                 0      F    сотрудник     0   
21510        женат / замужем                 0      F    сотрудник     0   

       total_income                           purpose  
12              NaN                   сыграть свадьбу  
26              NaN                       образование  
29              NaN  строительство жилой недвижимости  
41              NaN  сделка с подержанным автомобилем  
55              NaN                   сыграть свадьбу  
...             ...                               ...  
21489           NaN              сделка с автомобилем  
21495           NaN                           свадьба  
21497           NaN        строительство недвижимости  
21502           NaN  строительство жилой недвижимости  
21510           NaN           приобретение автомобиля  

[2174 rows x 12 columns]

**Поищем долю пропусков в столбце "total_income"**

In [7]:
data.loc[data["total_income"].isna()]["children"].count() / data.shape[0]

0.10099883855981417

### Задание 1

**опишите, какие пропущенные значения вы обнаружили:**

- *NaN*
  
**проверьте, какую долю составляют пропущенные значения в каждом из столбцов с пропусками:**

- *Приблизительно 10 процентов*
  
**приведите возможные причины появления пропусков в данных:**

- *Одна из причин: days_employed тоже NaN, то есть общий трудовой стаж NaN следовательно и доход NaN*

**объясните, почему заполнить пропуски медианным значением — лучшее решение для количественных переменных:**

- *Медианное значение не вносит существенных искажений(она не чувствительна к выбросам по сравнению заполнением со средним         значением)*

**Заполним проуски медианными значениями:**

In [8]:
data.loc[data["total_income"].isna(), "total_income"] = data["total_income"].median()

In [9]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      21525 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


Как видим пропусков в столбце зарплат больше нет

### Задание 2

**Причину отрицательности стажа не выявил, но скорее всего она связано со столбцом возраста либо при загрузке данных**

### Уберем все отрицательные значения (преобразуем в положительные с помощью abs):

In [10]:
data["days_employed"] = data["days_employed"].abs()

Теперь посмотрим на пропуски в столбце общего стажа и посмотрим с чем они связаны

In [11]:
data.loc[data["days_employed"].isna()]

children  days_employed  dob_years education  education_id  \
12            0            NaN         65   среднее             1   
26            0            NaN         41   среднее             1   
29            0            NaN         63   среднее             1   
41            0            NaN         50   среднее             1   
55            0            NaN         54   среднее             1   
...         ...            ...        ...       ...           ...   
21489         2            NaN         47   Среднее             1   
21495         1            NaN         50   среднее             1   
21497         0            NaN         48    ВЫСШЕЕ             0   
21502         1            NaN         42   среднее             1   
21510         2            NaN         28   среднее             1   

               family_status  family_status_id gender  income_type  debt  \
12          гражданский брак                 1      M    пенсионер     0   
26           женат / замужем                 0      M  госслужащий     0   
29     Не женат / не замужем                 4      F    пенсионер     0   
41           женат / замужем                 0      F  госслужащий     0   
55          гражданский брак                 1      F    пенсионер     1   
...                      ...               ...    ...          ...   ...   
21489        женат / замужем                 0      M    компаньон     0   
21495       гражданский брак                 1      F    сотрудник     0   
21497        женат / замужем                 0      F    компаньон     0   
21502        женат / замужем                 0      F    сотрудник     0   
21510        женат / замужем                 0      F    сотрудник     0   

        total_income                           purpose  
12     145017.937533                   сыграть свадьбу  
26     145017.937533                       образование  
29     145017.937533  строительство жилой недвижимости  
41     145017.937533  сделка с подержанным автомобилем  
55     145017.937533                   сыграть свадьбу  
...              ...                               ...  
21489  145017.937533              сделка с автомобилем  
21495  145017.937533                           свадьба  
21497  145017.937533        строительство недвижимости  
21502  145017.937533  строительство жилой недвижимости  
21510  145017.937533           приобретение автомобиля  

[2174 rows x 12 columns]

Непонятно с чем связаны с пропуски, но зато столкнулись с другой проблемой:

*По идее максимальный стаж приблизительно 50 лет(18250 в днях)*

In [12]:
data.loc[data["days_employed"]>18250].sort_values(by = "days_employed")

children  days_employed  dob_years education  education_id  \
16335         1   18388.949901         61   среднее             1   
20444         0  328728.720605         72   среднее             1   
9328          2  328734.923996         41    высшее             0   
17782         0  328771.341387         56   среднее             1   
14783         0  328795.726728         62    высшее             0   
...         ...            ...        ...       ...           ...   
7794          0  401663.850046         61   среднее             1   
2156          0  401674.466633         60   среднее             1   
7664          1  401675.093434         61   среднее             1   
10006         0  401715.811749         69    высшее             0   
6954          0  401755.400475         56   среднее             1   

               family_status  family_status_id gender income_type  debt  \
16335        женат / замужем                 0      F   сотрудник     0   
20444         вдовец / вдова                 2      F   пенсионер     0   
9328         женат / замужем                 0      M   пенсионер     0   
17782        женат / замужем                 0      F   пенсионер     0   
14783        женат / замужем                 0      F   пенсионер     0   
...                      ...               ...    ...         ...   ...   
7794        гражданский брак                 1      F   пенсионер     0   
2156         женат / замужем                 0      M   пенсионер     0   
7664         женат / замужем                 0      F   пенсионер     0   
10006  Не женат / не замужем                 4      F   пенсионер     0   
6954          вдовец / вдова                 2      F   пенсионер     0   

        total_income                                purpose  
16335  186178.934089               операции с недвижимостью  
20444   96519.339647                покупка жилья для семьи  
9328   126997.497760        операции со своей недвижимостью  
17782   68648.047062  операции с коммерческой недвижимостью  
14783   79940.196752           на покупку своего автомобиля  
...              ...                                    ...  
7794    48286.441362                                свадьба  
2156   325395.724541                             автомобили  
7664   126214.519212                      операции с жильем  
10006   57390.256908                  получение образования  
6954   176278.441171                           ремонт жилью  

[3446 rows x 12 columns]

#### То есть у нас есть пропуски в этом столбце, и еще некорректные данные
(между 18388 и 328728.720605 огромный разрыв, стаж не может быть = 328728.720605, это **900 лет**)

In [13]:
data.loc[data["days_employed"]>18250].sort_values(by = "days_employed")

children  days_employed  dob_years education  education_id  \
16335         1   18388.949901         61   среднее             1   
20444         0  328728.720605         72   среднее             1   
9328          2  328734.923996         41    высшее             0   
17782         0  328771.341387         56   среднее             1   
14783         0  328795.726728         62    высшее             0   
...         ...            ...        ...       ...           ...   
7794          0  401663.850046         61   среднее             1   
2156          0  401674.466633         60   среднее             1   
7664          1  401675.093434         61   среднее             1   
10006         0  401715.811749         69    высшее             0   
6954          0  401755.400475         56   среднее             1   

               family_status  family_status_id gender income_type  debt  \
16335        женат / замужем                 0      F   сотрудник     0   
20444         вдовец / вдова                 2      F   пенсионер     0   
9328         женат / замужем                 0      M   пенсионер     0   
17782        женат / замужем                 0      F   пенсионер     0   
14783        женат / замужем                 0      F   пенсионер     0   
...                      ...               ...    ...         ...   ...   
7794        гражданский брак                 1      F   пенсионер     0   
2156         женат / замужем                 0      M   пенсионер     0   
7664         женат / замужем                 0      F   пенсионер     0   
10006  Не женат / не замужем                 4      F   пенсионер     0   
6954          вдовец / вдова                 2      F   пенсионер     0   

        total_income                                purpose  
16335  186178.934089               операции с недвижимостью  
20444   96519.339647                покупка жилья для семьи  
9328   126997.497760        операции со своей недвижимостью  
17782   68648.047062  операции с коммерческой недвижимостью  
14783   79940.196752           на покупку своего автомобиля  
...              ...                                    ...  
7794    48286.441362                                свадьба  
2156   325395.724541                             автомобили  
7664   126214.519212                      операции с жильем  
10006   57390.256908                  получение образования  
6954   176278.441171                           ремонт жилью  

[3446 rows x 12 columns]





**Для начала заполним пропуски в days_employed медианными значениями по этому столбцу**

In [14]:
data.loc[data["days_employed"].isna(), "days_employed"] = data["days_employed"].median()

*На счет огромного стажа: его лучше оставить, не думаю что он на что-то повлияет. Можно изменить на медианное значение, но тогда будет так, что пенсионер будет со стажом 7 лет (все строки с огромным стажом принадлежат песионерам)*

**Аномалии еще есть в столбце дети** 

Посмотрим через unique

In [15]:
data["children"].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5], dtype=int64)

Выведем кол-во людей с отрицательным кол-ом детей

In [16]:
data["children"].value_counts()

 0     14149
 1      4818
 2      2055
 3       330
 20       76
-1        47
 4        41
 5         9
Name: children, dtype: int64

Целых **47** человек с -1 детьми

Теперь выведем их:

In [17]:
data.loc[data["children"]== -1].head(15)

children  days_employed  dob_years education  education_id  \
291         -1    4417.703588         46   среднее             1   
705         -1     902.084528         50   среднее             1   
742         -1    3174.456205         57   среднее             1   
800         -1  349987.852217         54   среднее             1   
941         -1    2194.220567         57   Среднее             1   
1363        -1    1195.264956         55   СРЕДНЕЕ             1   
1929        -1    1461.303336         38   среднее             1   
2073        -1    2539.761232         42   среднее             1   
3814        -1    3045.290443         26   Среднее             1   
4201        -1     901.101738         41   среднее             1   
4402        -1  398001.302888         64   СРЕДНЕЕ             1   
4542        -1    1811.899756         32   среднее             1   
5272        -1    1143.485347         46   среднее             1   
6013        -1    1361.258696         46    высшее             0   
6385        -1  370215.476226         48   среднее             1   

              family_status  family_status_id gender  income_type  debt  \
291        гражданский брак                 1      F    сотрудник     0   
705         женат / замужем                 0      F  госслужащий     0   
742         женат / замужем                 0      F    сотрудник     0   
800   Не женат / не замужем                 4      F    пенсионер     0   
941         женат / замужем                 0      F    пенсионер     0   
1363        женат / замужем                 0      F    компаньон     0   
1929  Не женат / не замужем                 4      M    сотрудник     0   
2073              в разводе                 3      F    компаньон     0   
3814       гражданский брак                 1      F  госслужащий     0   
4201        женат / замужем                 0      F  госслужащий     0   
4402        женат / замужем                 0      F    пенсионер     0   
4542        женат / замужем                 0      F    сотрудник     0   
5272         вдовец / вдова                 2      F    сотрудник     0   
6013        женат / замужем                 0      F    сотрудник     0   
6385         вдовец / вдова                 2      F    пенсионер     0   

       total_income                                 purpose  
291   102816.346412                  профильное образование  
705   137882.899271                 приобретение автомобиля  
742    64268.044444              дополнительное образование  
800    86293.724153              дополнительное образование  
941   145017.937533            на покупку своего автомобиля  
1363   69550.699692                  профильное образование  
1929  109121.569013                           покупка жилья  
2073  162638.609373                           покупка жилья  
3814  131892.785435                   на проведение свадьбы  
4201  226375.766751         операции со своей недвижимостью  
4402  163264.062064                    покупка недвижимости  
4542  160544.718446                             образование  
5272  278708.018820                    покупка недвижимости  
6013  143008.454914  строительство собственной недвижимости  
6385   36052.447435                             образование

*Посмотрев таблицу и не найдя ничего общего, я, будущи студентом, огорчился* (ну у некоторых совпадают ошибки со столбцом days_employed)

**Поэтому разумном решением будет изменить -1 на 0**

In [18]:
data.loc[data["children"]== -1, "children"] = 0

Проверим:

In [19]:
data["children"].value_counts()

0     14196
1      4818
2      2055
3       330
20       76
4        41
5         9
Name: children, dtype: int64

### Во втором задании избавились от некоторых аномалии, некторые оставили и заполнили пропуски в days_employed медианными значениями по этому столбцу. И еще *попытались* дать обоснование аномалиям 

### Задание 3

**Заменим вещественный тип столбца total_income на целочисленный с помощью метода astype()**

In [20]:
data["total_income"] = data["total_income"].astype(int)

In [21]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     21525 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      21525 non-null  int32  
 11  purpose           21525 non-null  object 
dtypes: float64(1), int32(1), int64(5), object(5)
memory usage: 1.9+ MB


*Как мы видим столбец total_income теперь иеет целочисленный тип*

### Задание 4

Для начала найдем строки-дубликаты

In [22]:
data

children  days_employed  dob_years education  education_id  \
0             1    8437.673028         42    высшее             0   
1             1    4024.803754         36   среднее             1   
2             0    5623.422610         33   Среднее             1   
3             3    4124.747207         32   среднее             1   
4             0  340266.072047         53   среднее             1   
...         ...            ...        ...       ...           ...   
21520         1    4529.316663         43   среднее             1   
21521         0  343937.404131         67   среднее             1   
21522         1    2113.346888         38   среднее             1   
21523         3    3112.481705         38   среднее             1   
21524         2    1984.507589         40   среднее             1   

          family_status  family_status_id gender income_type  debt  \
0       женат / замужем                 0      F   сотрудник     0   
1       женат / замужем                 0      F   сотрудник     0   
2       женат / замужем                 0      M   сотрудник     0   
3       женат / замужем                 0      M   сотрудник     0   
4      гражданский брак                 1      F   пенсионер     0   
...                 ...               ...    ...         ...   ...   
21520  гражданский брак                 1      F   компаньон     0   
21521   женат / замужем                 0      F   пенсионер     0   
21522  гражданский брак                 1      M   сотрудник     1   
21523   женат / замужем                 0      M   сотрудник     1   
21524   женат / замужем                 0      F   сотрудник     0   

       total_income                       purpose  
0            253875                 покупка жилья  
1            112080       приобретение автомобиля  
2            145885                 покупка жилья  
3            267628    дополнительное образование  
4            158616               сыграть свадьбу  
...             ...                           ...  
21520        224791             операции с жильем  
21521        155999          сделка с автомобилем  
21522         89672                  недвижимость  
21523        244093  на покупку своего автомобиля  
21524         82047         на покупку автомобиля  

[21525 rows x 12 columns]

In [23]:
data["education"].unique()

array(['высшее', 'среднее', 'Среднее', 'СРЕДНЕЕ', 'ВЫСШЕЕ',
       'неоконченное высшее', 'начальное', 'Высшее',
       'НЕОКОНЧЕННОЕ ВЫСШЕЕ', 'Неоконченное высшее', 'НАЧАЛЬНОЕ',
       'Начальное', 'Ученая степень', 'УЧЕНАЯ СТЕПЕНЬ', 'ученая степень'],
      dtype=object)

Слова в разных регистрах, так не пойдет, исправим это

In [24]:
data["education"] = data["education"].str.lower()

Привели все слова к нижнему регистру, теперь проверим:

In [25]:
data["education"].unique()

array(['высшее', 'среднее', 'неоконченное высшее', 'начальное',
       'ученая степень'], dtype=object)

Поищем явные дубликаты

In [26]:
len(data.duplicated()) == data.shape[0]

True

Тут я сравниваю кол-во строк data и кол-во не дубликатов. Вернулось True, значит явных дубликатов нет

Поищем еще неявные дубликаты

In [27]:
data.head()

children  days_employed  dob_years education  education_id  \
0         1    8437.673028         42    высшее             0   
1         1    4024.803754         36   среднее             1   
2         0    5623.422610         33   среднее             1   
3         3    4124.747207         32   среднее             1   
4         0  340266.072047         53   среднее             1   

      family_status  family_status_id gender income_type  debt  total_income  \
0   женат / замужем                 0      F   сотрудник     0        253875   
1   женат / замужем                 0      F   сотрудник     0        112080   
2   женат / замужем                 0      M   сотрудник     0        145885   
3   женат / замужем                 0      M   сотрудник     0        267628   
4  гражданский брак                 1      F   пенсионер     0        158616   

                      purpose  
0               покупка жилья  
1     приобретение автомобиля  
2               покупка жилья  
3  дополнительное образование  
4             сыграть свадьбу

In [28]:
d_purpose = data["purpose"].unique()

In [29]:
sorted_unq_purpose = sorted(d_purpose)

In [30]:
sorted_unq_purpose

['автомобили',
 'автомобиль',
 'высшее образование',
 'дополнительное образование',
 'жилье',
 'заняться высшим образованием',
 'заняться образованием',
 'на покупку автомобиля',
 'на покупку подержанного автомобиля',
 'на покупку своего автомобиля',
 'на проведение свадьбы',
 'недвижимость',
 'образование',
 'операции с жильем',
 'операции с коммерческой недвижимостью',
 'операции с недвижимостью',
 'операции со своей недвижимостью',
 'покупка жилой недвижимости',
 'покупка жилья',
 'покупка жилья для сдачи',
 'покупка жилья для семьи',
 'покупка коммерческой недвижимости',
 'покупка недвижимости',
 'покупка своего жилья',
 'получение высшего образования',
 'получение дополнительного образования',
 'получение образования',
 'приобретение автомобиля',
 'профильное образование',
 'ремонт жилью',
 'свадьба',
 'свой автомобиль',
 'сделка с автомобилем',
 'сделка с подержанным автомобилем',
 'строительство жилой недвижимости',
 'строительство недвижимости',
 'строительство собственной недв

Тут, конечно, можно объединить некоторые строки вместе (например: строки связанные со свадьбой, образованием, автомобилем) но это выполним позже

**Явные и неявные дубликаты появляются из-за различных проблем, например:**

* Ошибки при сборе или вводе данных: Если данные были введены несколько раз или скопированы, это может привести к созданию явных дубликатов.

* Несовместимость регистра: Если значения в столбце содержат разный регистр (например, "apple" и "Apple"), то при сравнении строк они будут рассматриваться как разные, что может привести к неявным дубликатам.

* Ошибки при сохранении данных: Если данные были сохранены несколько раз, это может привести к созданию явных дубликатов.

### Задание 5

In [31]:
data.head()

children  days_employed  dob_years education  education_id  \
0         1    8437.673028         42    высшее             0   
1         1    4024.803754         36   среднее             1   
2         0    5623.422610         33   среднее             1   
3         3    4124.747207         32   среднее             1   
4         0  340266.072047         53   среднее             1   

      family_status  family_status_id gender income_type  debt  total_income  \
0   женат / замужем                 0      F   сотрудник     0        253875   
1   женат / замужем                 0      F   сотрудник     0        112080   
2   женат / замужем                 0      M   сотрудник     0        145885   
3   женат / замужем                 0      M   сотрудник     0        267628   
4  гражданский брак                 1      F   пенсионер     0        158616   

                      purpose  
0               покупка жилья  
1     приобретение автомобиля  
2               покупка жилья  
3  дополнительное образование  
4             сыграть свадьбу

Создадим 2 новых датафрейма. Первым создадим датафрайм образования

In [32]:
data["education_id"].unique()

array([0, 1, 2, 3, 4], dtype=int64)

In [33]:
data["education"].unique()

array(['высшее', 'среднее', 'неоконченное высшее', 'начальное',
       'ученая степень'], dtype=object)

In [34]:
columns_ed = ["education_id", "education"]

In [35]:
data_ed = [
    [0, 'высшее'],
    [1, 'среднее'],
    [2, 'неоконченное высшее'],
    [3, 'начальное'],
    [4, 'ученая степень']
      ]

In [36]:
education = pd.DataFrame(data = data_ed, columns = columns_ed)

In [37]:
education

education_id            education
0             0               высшее
1             1              среднее
2             2  неоконченное высшее
3             3            начальное
4             4       ученая степень

Теперь создадим датафрайм семьи

In [38]:
data.head()

children  days_employed  dob_years education  education_id  \
0         1    8437.673028         42    высшее             0   
1         1    4024.803754         36   среднее             1   
2         0    5623.422610         33   среднее             1   
3         3    4124.747207         32   среднее             1   
4         0  340266.072047         53   среднее             1   

      family_status  family_status_id gender income_type  debt  total_income  \
0   женат / замужем                 0      F   сотрудник     0        253875   
1   женат / замужем                 0      F   сотрудник     0        112080   
2   женат / замужем                 0      M   сотрудник     0        145885   
3   женат / замужем                 0      M   сотрудник     0        267628   
4  гражданский брак                 1      F   пенсионер     0        158616   

                      purpose  
0               покупка жилья  
1     приобретение автомобиля  
2               покупка жилья  
3  дополнительное образование  
4             сыграть свадьбу

In [39]:
data["family_status_id"].unique()

array([0, 1, 2, 3, 4], dtype=int64)

In [40]:
data["family_status"].unique()

array(['женат / замужем', 'гражданский брак', 'вдовец / вдова',
       'в разводе', 'Не женат / не замужем'], dtype=object)

In [41]:
columns_fam = ["family_status_id", "family_status"]

In [42]:
data_fam = [
    [0, 'женат / замужем'],
    [1, 'гражданский брак'],
    [2, 'вдовец / вдова'],
    [3, 'в разводе'],
    [4, 'Не женат / не замужем']
      ]

In [43]:
family = pd.DataFrame(data = data_fam, columns = columns_fam)

In [44]:
family

family_status_id          family_status
0                 0        женат / замужем
1                 1       гражданский брак
2                 2         вдовец / вдова
3                 3              в разводе
4                 4  Не женат / не замужем

Теперь удалим из исходного датафрейма столбцы *education* и *family_status*:

In [45]:
data = data.drop(columns = ["education", "family_status"])

In [46]:
data

children  days_employed  dob_years  education_id  family_status_id  \
0             1    8437.673028         42             0                 0   
1             1    4024.803754         36             1                 0   
2             0    5623.422610         33             1                 0   
3             3    4124.747207         32             1                 0   
4             0  340266.072047         53             1                 1   
...         ...            ...        ...           ...               ...   
21520         1    4529.316663         43             1                 1   
21521         0  343937.404131         67             1                 0   
21522         1    2113.346888         38             1                 1   
21523         3    3112.481705         38             1                 0   
21524         2    1984.507589         40             1                 0   

      gender income_type  debt  total_income                       purpose  
0          F   сотрудник     0        253875                 покупка жилья  
1          F   сотрудник     0        112080       приобретение автомобиля  
2          M   сотрудник     0        145885                 покупка жилья  
3          M   сотрудник     0        267628    дополнительное образование  
4          F   пенсионер     0        158616               сыграть свадьбу  
...      ...         ...   ...           ...                           ...  
21520      F   компаньон     0        224791             операции с жильем  
21521      F   пенсионер     0        155999          сделка с автомобилем  
21522      M   сотрудник     1         89672                  недвижимость  
21523      M   сотрудник     1        244093  на покупку своего автомобиля  
21524      F   сотрудник     0         82047         на покупку автомобиля  

[21525 rows x 10 columns]

### Задание 6

создадим столбец "total_income_category"

In [47]:
def categorize_income(total_income):
    if total_income <= 30000:
        return "E"
    elif total_income <= 50000:
        return "D"
    elif total_income <= 200000:
        return "C"
    elif total_income <= 1000000:
        return "B"
    else:
        return "A"

In [48]:
data["total_income_category"] = data["total_income"].apply(categorize_income)

Здесь мы создали функцию *categorize_income* который принимает доход и возращает значение исходя от уровня дохода. Затем применили метод *apply*, чтобы заполнить столбец *total_income_category*

In [49]:
data.head()

children  days_employed  dob_years  education_id  family_status_id gender  \
0         1    8437.673028         42             0                 0      F   
1         1    4024.803754         36             1                 0      F   
2         0    5623.422610         33             1                 0      M   
3         3    4124.747207         32             1                 0      M   
4         0  340266.072047         53             1                 1      F   

  income_type  debt  total_income                     purpose  \
0   сотрудник     0        253875               покупка жилья   
1   сотрудник     0        112080     приобретение автомобиля   
2   сотрудник     0        145885               покупка жилья   
3   сотрудник     0        267628  дополнительное образование   
4   пенсионер     0        158616             сыграть свадьбу   

  total_income_category  
0                     B  
1                     C  
2                     C  
3                     B  
4                     C

### Задание 7

Ранее мы создавали переменную, которая выводит отсортированные уникальные значения из столбца purpose

In [50]:
sorted_unq_purpose

['автомобили',
 'автомобиль',
 'высшее образование',
 'дополнительное образование',
 'жилье',
 'заняться высшим образованием',
 'заняться образованием',
 'на покупку автомобиля',
 'на покупку подержанного автомобиля',
 'на покупку своего автомобиля',
 'на проведение свадьбы',
 'недвижимость',
 'образование',
 'операции с жильем',
 'операции с коммерческой недвижимостью',
 'операции с недвижимостью',
 'операции со своей недвижимостью',
 'покупка жилой недвижимости',
 'покупка жилья',
 'покупка жилья для сдачи',
 'покупка жилья для семьи',
 'покупка коммерческой недвижимости',
 'покупка недвижимости',
 'покупка своего жилья',
 'получение высшего образования',
 'получение дополнительного образования',
 'получение образования',
 'приобретение автомобиля',
 'профильное образование',
 'ремонт жилью',
 'свадьба',
 'свой автомобиль',
 'сделка с автомобилем',
 'сделка с подержанным автомобилем',
 'строительство жилой недвижимости',
 'строительство недвижимости',
 'строительство собственной недв

In [51]:
def categorize_purpose(purpose):
    if "авто" in purpose:
        return 'операции с автомобилем'
    
    elif "образован" in purpose:
        return 'получение образования'
    
    elif "свадьб" in purpose:
        return 'проведение свадьбы'
    
    else:
        return 'операции с недвижимостью'

In [52]:
data["purpose_category"] = data["purpose"].apply(categorize_purpose)

In [53]:
data["purpose_category"].value_counts()

операции с недвижимостью    10840
операции с автомобилем       4315
получение образования        4022
проведение свадьбы           2348
Name: purpose_category, dtype: int64

Больше половины людей берут кредит для операции с недвижимостью

# Шаг 3

*Есть ли зависимость между количеством детей и возвратом кредита в срок?*

In [54]:
data.head(5)

children  days_employed  dob_years  education_id  family_status_id gender  \
0         1    8437.673028         42             0                 0      F   
1         1    4024.803754         36             1                 0      F   
2         0    5623.422610         33             1                 0      M   
3         3    4124.747207         32             1                 0      M   
4         0  340266.072047         53             1                 1      F   

  income_type  debt  total_income                     purpose  \
0   сотрудник     0        253875               покупка жилья   
1   сотрудник     0        112080     приобретение автомобиля   
2   сотрудник     0        145885               покупка жилья   
3   сотрудник     0        267628  дополнительное образование   
4   пенсионер     0        158616             сыграть свадьбу   

  total_income_category          purpose_category  
0                     B  операции с недвижимостью  
1                     C    операции с автомобилем  
2                     C  операции с недвижимостью  
3                     B     получение образования  
4                     C        проведение свадьбы

**Для ответа на вопрос создадим сводную таблицу и отсортируем по возрастанию среднего значения**

In [55]:
result = data.pivot_table(index="children", values="debt", aggfunc=["sum", "count", "mean"])

result.columns = result.columns.droplevel(1)

result = result.sort_values(by="mean")

print(result)

           sum  count      mean
children                       
5            0      9  0.000000
0         1064  14196  0.074951
3           27    330  0.081818
1          444   4818  0.092154
2          194   2055  0.094404
4            4     41  0.097561
20           8     76  0.105263


Людей с кол-ом детей 5, 4 и 20 можем не учитывать, так как таких людей слишком мало.

Можно сделать вывод о том что, действительно **есть зависимость между количеством детей и возвратом кредита в срок**

- Люди у которых нет детей: не возращают 7,5% 
- У которых 3 ребенка: не возращают 8,2%
- У которых 1 ребенок: не возращают 9,2%
- У которых 2 ребенка: не возращают 9,4%

*Есть ли зависимость между семейным положением и возвратом кредита в срок?*

In [56]:
family

family_status_id          family_status
0                 0        женат / замужем
1                 1       гражданский брак
2                 2         вдовец / вдова
3                 3              в разводе
4                 4  Не женат / не замужем

In [57]:
result_fam = data.pivot_table(index="family_status_id", values="debt", aggfunc=["sum", "count", "mean"])

result_fam.columns = result_fam.columns.droplevel(1)

result_fam = result_fam.sort_values(by="mean")

# Тут поменяем id (0,1,2...) на наименования 
b = family.set_index('family_status_id')
result_fam = result_fam.rename(index=b['family_status'])

result_fam

sum  count      mean
family_status_id                           
вдовец / вдова          63    960  0.065625
в разводе               85   1195  0.071130
женат / замужем        931  12380  0.075202
гражданский брак       388   4177  0.092890
Не женат / не замужем  274   2813  0.097405

Можно сделать вывод о том что, между семейным положением и возвратом кредита в срок нет огромной зависимости:
Из людей которые имеют семейное положение *вдовец / вдова* (максимальный процент возврата)  не возращают 6,57%, а людей у которых  *Не женат / не замужем* (минимальный процент возрата) не возращают 9,74%. То есть нет огромного разброса между этими значениями

*Есть ли зависимость между уровнем дохода и возвратом кредита в срок?*

In [58]:
data.head()

children  days_employed  dob_years  education_id  family_status_id gender  \
0         1    8437.673028         42             0                 0      F   
1         1    4024.803754         36             1                 0      F   
2         0    5623.422610         33             1                 0      M   
3         3    4124.747207         32             1                 0      M   
4         0  340266.072047         53             1                 1      F   

  income_type  debt  total_income                     purpose  \
0   сотрудник     0        253875               покупка жилья   
1   сотрудник     0        112080     приобретение автомобиля   
2   сотрудник     0        145885               покупка жилья   
3   сотрудник     0        267628  дополнительное образование   
4   пенсионер     0        158616             сыграть свадьбу   

  total_income_category          purpose_category  
0                     B  операции с недвижимостью  
1                     C    операции с автомобилем  
2                     C  операции с недвижимостью  
3                     B     получение образования  
4                     C        проведение свадьбы

Повторим те же действия, только со столбцом *total_income_category*

In [61]:
result_income = data.pivot_table(index="total_income_category", values="debt", aggfunc=["sum", "count", "mean"])

result_income.columns = result_income.columns.droplevel(1)

result_income = result_income.sort_values(by="mean")

result_income

sum  count      mean
total_income_category                       
D                        21    350  0.060000
B                       356   5041  0.070621
A                         2     25  0.080000
C                      1360  16087  0.084540
E                         2     22  0.090909

То, что в конце расположилась категория "E" не удивляет (0–30000 — 'E') 

Но из-за того что в категоряих A и E мало данных их можно отбросить

Получается ситуация такая же, как и в семейном положении, максимальный процент не возрата у категории "C" = 8,45% , а минимальный у "D" = 6%. Получается как таковой зависимости нет, но если отбросить столбец "D"(допустим там тоже мало значений), то получим что люди с более высокой зарплатой (B) имеют наименьший шанс невозрата кредита в срок нежели "C"

*Как разные цели кредита влияют на его возврат в срок?*

In [62]:
result_purpose = data.pivot_table(index="purpose_category", values="debt", aggfunc=["sum", "count", "mean"])

result_purpose.columns = result_purpose.columns.droplevel(1)

result_purpose = result_purpose.sort_values(by="mean")

result_purpose

sum  count      mean
purpose_category                              
операции с недвижимостью  782  10840  0.072140
проведение свадьбы        186   2348  0.079216
получение образования     370   4022  0.091994
операции с автомобилем    403   4315  0.093395

Тут не будем отбрасывать данные.

Можно сделать вывод, что разные цели кредита могут влиять на его возврат в срок. В данном случае, кредиты на *операции с недвижимостью* имеют **наименьший** процент невозврата кредита (7,2%), в то время как кредиты на *операции с автомобилем* имеют **наивысший** процент невозврата (9,3%). Кредиты на *проведение свадьбы* и *получение образования* также имеют высокий процент невозврата (7,9% и 9,1% соответственно).

# Шаг 4

В этом проекте разобрались, влияет ли семейное положение, количество детей, уровень зарплаты, цели клиента на факт погашения кредита в срок (и обошлись без использования  try...catch)

- **Семейное положение**: есть малая зависимость - минимальный невозврат у людей с семейным положением "вдовец / вдова", а максимальный у "Не женат / не замужем"
- **Количество детей**: определенно есть зависимость, чем больше детей тем больше процент невозврата
- **Уровень зарплаты**: если пренебречь малыми данными, то получим, что люди с более высокой зарплатой имеют наименьший шанс невозрата кредита в срок, а люди у которых низкая зарплата имеют наивысший шанс невозрата кредита в срок
- **Цели**: *операции с недвижимостью* имеют наименьший процент невозврата кредита, в то время как кредиты на *операции с автомобилем* имеют наивысший процент невозврата